src/metatensor/models/experimental/soap_bpnn/trainer.py

```diff
@@ -0,0 +1,285 @@
+import logging
+import warnings
+from pathlib import Path
+from typing import List, Union
+
+import torch
+from metatensor.learn.data import DataLoader
+
+from ...utils.composition import calculate_composition_weights
+from ...utils.data import (
+    CombinedDataLoader,
+    Dataset,
+    collate_fn,
+    get_all_targets,
+    get_all_types,
+)
+from ...utils.evaluate_model import evaluate_model
+from ...utils.extract_targets import get_outputs_dict
+from ...utils.logging import MetricLogger
+from ...utils.loss import TensorMapDictLoss
+from ...utils.metrics import RMSEAccumulator
+from ...utils.per_atom import divide_by_num_atoms
+from .model import SOAPBPNN
+
+
+logger = logging.getLogger(__name__)
+
+
+# Filter out the second derivative and device warnings from rascaline-torch
+warnings.filterwarnings("ignore", category=UserWarning, message="second derivative")
+warnings.filterwarnings(
+    "ignore", category=UserWarning, message="Systems data is on device"
+)
+
+
+class Trainer:
+    def __init__(self, train_hypers):
+        self.hypers = train_hypers
+
+    def train(
+        self,
+        model: SOAPBPNN,
+        devices: List[torch.device],
+        train_datasets: List[Union[Dataset, torch.utils.data.Subset]],
+        validation_datasets: List[Union[Dataset, torch.utils.data.Subset]],
+        checkpoints_dir: str,
+    ):
+        dtype = train_datasets[0][0]["system"].positions.dtype
+
+        # only one device, as we don't support multi-gpu for now
+        assert len(devices) == 1
+        device = devices[0]
+
+        logger.info(f"training on device {device} with dtype {dtype}")
+        model.to(device=device, dtype=dtype)
+
+        # Calculate and set the composition weights for all targets:
+        logger.info("Calculating composition weights")
+        for target_name in model.outputs.keys():
+            if "mtm::aux::" in target_name:
+                continue
+            # TODO: warn in the documentation that capabilities that are already
+            # present in the model won't recalculate the composition weights
+            # find the datasets that contain the target:
+
+            if target_name in self.hypers["fixed_composition_weights"].keys():
+                logger.info(
+                    f"For {target_name}, model will proceed with "
+                    "user-supplied composition weights"
+                )
+                cur_weight_dict = self.hypers["fixed_composition_weights"][target_name]
+                species = []
+                num_species = len(cur_weight_dict)
+                fixed_weights = torch.zeros(num_species, dtype=dtype, device=device)
+
+                for ii, (key, weight) in enumerate(cur_weight_dict.items()):
+                    species.append(key)
+                    fixed_weights[ii] = weight
+
+                all_types = []
+                for dataset in train_datasets:
+                    all_types += get_all_types(dataset)
```

Comment:
all_types should be contained in DatasetInfo

Reply:
Good catch!